mediagrains/utils/h264_parser.py

```diff
@@ -0,0 +1,700 @@
+        frame_info.height = frame_info.coded_height - crop_unit_y * sps.frame_crop_bottom_offset - y_offset
+
+        frame_info.sample_aspect_ratio = self._get_sample_aspect_ratio(sps)
+
+        if sps.timing_info_present_flag:
+            if sps.time_scale == 0 or sps.num_units_in_tick == 0:
+                raise ValueError("Invalid timing info: time_scale={}, num_units_in_tick={}".format(
+                    sps.time_scale, sps.num_units_in_tick
+                ))
+
+            # TODO: implement D.2.2 of H.264 spec. and explanation surrounding Table E-6
+            delta_tfi_divisor = 2
+            if primary_slice_headers[0].field_pic_flag:
+                delta_tfi_divisor = 1
+
+            frame_info.frame_rate = Fraction(sps.time_scale, sps.num_units_in_tick * delta_tfi_divisor)
+            frame_info.fixed_frame_rate = sps.fixed_frame_rate_flag != 0
+
+        # If all primary slices are an I type then this is a key frame
+        for slice_header in primary_slice_headers:
+            frame_info.key_frame = slice_header.slice_type in [2, 4, 7, 9]  # I, SI, I2, SI2
+            if not frame_info.key_frame:
+                break
+
+        return frame_info
+
+    def _get_sample_aspect_ratio(self, sps: SPS) -> Optional[Fraction]:
+        """Return the sample aspect ratio"""
+        aspect_ratios = [
+            (0, 1),
+            (1, 1),
+            (12, 11),
+            (10, 11),
+            (16, 11),
+            (40, 33),
+            (24, 11),
+            (20, 11),
+            (32, 11),
+            (80, 33),
+            (18, 11),
+            (15, 11),
+            (64, 33),
+            (160, 99),
+            (4, 3),
+            (3, 2),
+            (2, 1),
+        ]
+
+        if sps.aspect_ratio_idc == 255:  # extended SAR
+            return Fraction(sps.sar_width, sps.sar_height)
+        elif sps.aspect_ratio_idc < len(aspect_ratios):
+            return Fraction(aspect_ratios[sps.aspect_ratio_idc][0], aspect_ratios[sps.aspect_ratio_idc][1])
+        else:
+            return None
+
+    def _parse_nal_unit_type(self, bits: BitStream) -> int:
+        """Peek NALU for NAL Unit Type
+
+        :param bits: NALU bits
+        :returns: NAL Unit Type
+        """
+        self._parse_start_code_prefix(bits)
+
+        bits.read('uint:1')  # forbidden_zero_bit
+        bits.read('uint:2')  # nal_ref_idc
+        nal_unit_type = bits.read('uint:5')
+
+        return nal_unit_type
+
+    def _parse_sps(self, rbsp_bits: BitStream) -> SPS:
+        """Parse SPS NALU
+
+        :param bits: RBSP bits
+        :returns: parsed SPS
+        :raises ValueError: if a parse failure occurs
+        """
+        sps = SPS()
+
+        rbsp_bits.read('uint:1')  # forbidden_zero_bit
+        rbsp_bits.read('uint:2')  # nal_ref_idc
+        rbsp_bits.read('uint:5')  # nal_unit_type
+
+        sps.profile_idc = rbsp_bits.read('uint:8')
```

Comment:
The following section is quite hard to read. Could you spring some whitespace through it to help split it out a bit?

Reply:
It is intended to be read alongside the syntax definitions in the H.264 spec., i.e. the code is not intended to be understood but rather intended to easily compared with the text in the spec..